client/protocols/export.py

```diff
@@ -0,0 +1,145 @@
+                            data['assignid'])
+                        data['current'] = i + 1
+                    except Exception as e:
+                        data['current'] = i
+                        pickle.dump(data, open("export_cache.pkl", "wb"))
+                        dl_left = len(data['students']) - i
+                        print("Download failed. Run command again to continue.")
+                        print("{0} submissions left to download".format(dl_left))
+                        sys.exit(0)
+                        
+            except KeyboardInterrupt:
+                pickle.dump(data, open("export_cache.pkl", "wb"))
+                dl_left = len(data['students']) - i
+                print("Download interrupted. Run command again to continue.")
+                print("{0} submissions left to download".format(dl_left))
+                sys.exit(0)
+                
+            print("Submissions downloaded.")
+            if os.path.exists('export_cache.pkl'):
+                os.remove('export_cache.pkl')
+            sys.exit(0)
+            
+    def get_course(self):
+        """Gets a course ID to search for"""
+        if self.args.course > 0:
+            return self.args.course
+        print("Loading course list...")
+        response = self.request('course')
+        num = 0
+        courses = response['data']['results']
+        if len(courses) == 1:
+            return courses[0]['id']
+        elif len(courses) == 0:
+            print("No courses found.")
+            sys.exit(0)
+        for course in courses:
+            name = course['display_name']
+            offering = course['offering']
+            inst = course['institution']
+            print("{0}: {1} {2} {3}".format(num, name, offering, inst))
+            num += 1
+        return courses[int(input("Course #: "))]['id']
+        
+    def download_submission(self, student, name, assign_id):
+        """Downloads a student's final submission for an assignment"""
+        params = {'assignment': assign_id}
+        response = self.request('user/{0}/final_submission'.format(student[0]),
+            params)
+        if not response['data']:
+            print("No final submission for {0}".format(student[0]))
+            return
+        raw_data = response['data']
+        contents = raw_data['submission']['backup']['messages']['file_contents']
+        timestamp = datetime.strptime(raw_data['submission']['backup']['server_time'], GAE_DATETIME_FORMAT)
+        subm_id = raw_data['submission']['backup']['id']
+        subm_dir = name + "/" + student[0]
+        os.makedirs(subm_dir)
+        for filename in contents:
+            if filename != 'submit':
+                with open(subm_dir + "/" + filename, "w+") as f:
+                    f.write(contents[filename])
+                    f.flush()
+                    os.fsync(f.fileno())
+
+        with open(subm_dir + "/info.py" , "w+") as f:
```

Comment:
Is this file used by OK at some later point? If not, is there a reason why it's a Python file?